# MNIST Naive Bayes Model Comparison with scikit-learn

## Project Overview

This project compares multiple scikit-learn Naive Bayes formulations on the same normalized MNIST representation.

The supplied source material notes that scikit-learn provides several Naive Bayes implementations based on different distributional assumptions. The implementations evaluated here are:

- `GaussianNB`
- `MultinomialNB`
- `ComplementNB`
- `BernoulliNB`

Several Bernoulli configurations are also compared by changing the binarization threshold and smoothing parameter.

### Technical Coverage

- Gaussian Naive Bayes
- Multinomial Naive Bayes
- Complement Naive Bayes
- Bernoulli Naive Bayes
- Mean-based binarization
- Fixed 0.5 binarization
- Bernoulli smoothing parameter comparison
- Accuracy and confusion-matrix analysis

Reference: https://scikit-learn.org/stable/modules/naive_bayes.html

## 1. Environment

In [1]:
import tensorflow as tf
import numpy as np
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.naive_bayes import GaussianNB, MultinomialNB, ComplementNB, BernoulliNB


## 2. MNIST Data Preparation

The same MNIST train/test split is normalized and flattened into 784-dimensional feature vectors so every estimator is evaluated on the same data representation.

In [2]:
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.mnist.load_data()
x_train = np.float32(x_train/255).reshape(x_train.shape[0],-1)
x_test = np.float32(x_test/255).reshape(x_test.shape[0],-1)


11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


## 3. Gaussian Naive Bayes

The first comparison uses `GaussianNB`.

In [3]:
model =  GaussianNB()
model.fit(x_train, y_train)
pred = model.predict(x_test)
accuracy = accuracy_score(y_test, pred)
print(f'Accuracy = {accuracy:.4}')
cm = confusion_matrix(y_test, pred)
print(cm)

Accuracy = 0.5558
[[ 870    0    3    5    2    5   31    1   35   28]
 [   0 1079    2    1    0    0   10    0   38    5]
 [  79   25  266   91    5    2  269    4  271   20]
 [  32   39    6  353    2    3   51    8  409  107]
 [  19    2    5    4  168    7   63    7  210  497]
 [  71   25    1   20    3   44   40    2  586  100]
 [  12   12    3    1    1    7  895    0   26    1]
 [   0   15    2   10    5    1    5  280   39  671]
 [  13   72    3    7    3   11   12    4  648  201]
 [   5    7    3    6    1    0    1   13   18  955]]


## 4. Multinomial Naive Bayes

The second model uses `MultinomialNB`.

In [4]:
model =  MultinomialNB()
model.fit(x_train, y_train)
pred = model.predict(x_test)
accuracy = accuracy_score(y_test, pred)
print(f'Accuracy = {accuracy:.4}')
cm = confusion_matrix(y_test, pred)
print(cm)

Accuracy = 0.8357
[[ 912    0    2    6    1    8   14    1   36    0]
 [   0 1061    5    9    0    1    6    0   52    1]
 [  15   11  858   24   10    3   32   12   66    1]
 [   4   11   34  852    1   19    7   14   41   27]
 [   2    2    6    0  731    0   25    1   38  177]
 [  23   12    6  108   18  586   17    6   79   37]
 [  18   13   17    1    8   24  859    0   18    0]
 [   1   21   11    5   18    0    1  861   41   69]
 [   6   26   13   55   14   27    8    9  776   40]
 [   6    8    3   11   65    9    1   18   27  861]]


## 5. Complement Naive Bayes

The third model uses `ComplementNB`.

In [5]:
model =  ComplementNB()
model.fit(x_train, y_train)
pred = model.predict(x_test)
accuracy = accuracy_score(y_test, pred)
print(f'Accuracy = {accuracy:.4}')
cm = confusion_matrix(y_test, pred)
print(cm)

Accuracy = 0.7286
[[ 896    2    0    4    0    3   59    3   13    0]
 [   0 1108    7    7    0    1    5    0    7    0]
 [  93   34  750   38    2    1   71   26   16    1]
 [  25   22   33  833    0    1   15   61    8   12]
 [  76   60   13    1  605    9   29   34    3  152]
 [  84   36   13  203    9  426   31   46   20   24]
 [ 112   17   16    0    1    8  803    0    0    1]
 [  12   52   11    5    3    0    1  918    5   21]
 [  81  183   21  129    8   58   18   66  370   40]
 [  42   34    9   16   28    0    1  300    2  577]]


## 6. Bernoulli Naive Bayes

The default `BernoulliNB` configuration establishes the Bernoulli baseline.

In [6]:
model =  BernoulliNB()
model.fit(x_train, y_train)
pred = model.predict(x_test)
accuracy = accuracy_score(y_test, pred)
print(f'Accuracy = {accuracy:.4}')
cm = confusion_matrix(y_test, pred)
print(cm)

Accuracy = 0.8413
[[ 887    0    4    7    2   41   16    1   22    0]
 [   0 1085   10    5    0    9    6    0   19    1]
 [  19    8  852   29   17    4   32   14   55    2]
 [   5   15   34  844    0   13    9   15   49   26]
 [   2    6    4    0  795    4   21    1   23  126]
 [  23   12    7  129   30  627   16    8   21   19]
 [  18   18   15    2   13   35  851    0    6    0]
 [   1   24   14    4   15    0    0  871   27   72]
 [  16   23   13   76   17   22    7    6  758   36]
 [   9   13    5    9   74    8    0   24   24  843]]


## 7. Bernoulli Naive Bayes with Mean-Based Binarization

The binarization threshold is set to the mean value of the normalized training pixels.

In [7]:
model =  BernoulliNB(binarize=np.mean(x_train))
model.fit(x_train, y_train)
pred = model.predict(x_test)
accuracy = accuracy_score(y_test, pred)
print(f'Accuracy = {accuracy:.4}')
cm = confusion_matrix(y_test, pred)
print(cm)

Accuracy = 0.8462
[[ 894    0    4    4    1   33   22    1   21    0]
 [   0 1085   10    4    0   11    5    0   20    0]
 [  17   12  850   31   20    4   27   15   53    3]
 [   3   15   35  851    0   14    8   16   47   21]
 [   1    7    4    0  807    3   19    2   16  123]
 [  22   10    5  122   22  636   18    9   26   22]
 [  14   16   14    2   12   34  860    0    6    0]
 [   1   29   16    5   15    0    1  867   27   67]
 [  15   23   12   74   13   22   12    6  762   35]
 [  10   13    6    9   69    7    0   23   22  850]]


## 8. Bernoulli Naive Bayes with a 0.5 Threshold

A fixed binarization threshold of 0.5 provides a direct comparison with the mean-based threshold.

In [8]:
model =  BernoulliNB(binarize=0.5)
model.fit(x_train, y_train)
pred = model.predict(x_test)
accuracy = accuracy_score(y_test, pred)
print(f'Accuracy = {accuracy:.4}')
cm = confusion_matrix(y_test, pred)
print(cm)

Accuracy = 0.8427
[[ 882    0    3    4    1   51   21    1   17    0]
 [   0 1086    6    5    0    9    4    0   25    0]
 [  17   14  840   32   21    5   28   16   57    2]
 [   4   17   36  839    1   30    7   13   44   19]
 [   2    7    4    0  799    2   16    2   12  138]
 [  16   11    5  106   24  655   18    8   24   25]
 [  17   14   15    1   14   37  853    0    7    0]
 [   2   33   17    3   17    0    0  864   25   67]
 [  10   28   12   68   16   27   10    6  762   35]
 [  12   13    6    8   64    9    0   29   21  847]]


## 9. Bernoulli Naive Bayes with Adjusted Smoothing

The final configuration combines mean-based binarization with `alpha=0.05`.

In [9]:
model =  BernoulliNB(alpha=0.05, binarize=np.mean(x_train))
model.fit(x_train, y_train)
pred = model.predict(x_test)
accuracy = accuracy_score(y_test, pred)
print(f'Accuracy = {accuracy:.4}')
cm = confusion_matrix(y_test, pred)
print(cm)

Accuracy = 0.8465
[[ 894    0    4    4    1   33   22    1   21    0]
 [   0 1084   10    4    0   11    5    0   21    0]
 [  17   12  851   31   20    4   26   15   53    3]
 [   3   15   35  852    0   14    8   16   46   21]
 [   1    7    5    0  806    3   19    2   16  123]
 [  22    9    5  122   21  638   18    9   26   22]
 [  13   16   15    2   12   34  860    0    6    0]
 [   1   29   16    5   15    0    1  867   27   67]
 [  15   23   12   74   13   22   12    6  762   35]
 [  10   13    6    9   69    8    0   24   19  851]]


## 10. Comparative Results

| Model / Configuration | Saved Test Accuracy |
|---|---:|
| GaussianNB | 0.5558 |
| MultinomialNB | 0.8357 |
| ComplementNB | 0.7286 |
| BernoulliNB | 0.8413 |
| BernoulliNB + mean binarization | 0.8462 |
| BernoulliNB + 0.5 binarization | 0.8427 |
| BernoulliNB + mean binarization + alpha=0.05 | **0.8465** |

The saved results show that the Bernoulli configurations provide the strongest performance among the supplied scikit-learn comparisons, with the adjusted-smoothing configuration reaching 0.8465 test accuracy.

## 11. Key Findings

- Different Naive Bayes formulations can behave very differently on the same feature representation.
- The supplied MNIST comparison produces substantially lower accuracy with GaussianNB than with the Bernoulli and Multinomial formulations.
- Bernoulli performance changes with the binarization threshold.
- Adjusting the Bernoulli smoothing parameter produces the strongest saved scikit-learn result in the supplied comparison.
- Confusion matrices reveal the class-level structure behind the overall accuracy values.